client.go

```diff
@@ -818,3 +779,109 @@ func (c *TunaSessionClient) removeClosedSessions() {
 		c.Unlock()
 	}
 }
+
+func (c *TunaSessionClient) requestPubAddrs(remoteAddr string, config *nkn.DialConfig) (*PubAddrs, error) {
+	ctx := context.Background()
+	var cancel context.CancelFunc
+	if config.DialTimeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(config.DialTimeout)*time.Millisecond)
+		defer cancel()
+	}
+
+	buf, err := json.Marshal(&Request{Action: "getPubAddr"})
+	if err != nil {
+		return nil, err
+	}
+
+	respChan, err := c.multiClient.Send(nkn.NewStringArray(remoteAddr), buf, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var msg *nkn.Message
+	select {
+	case msg = <-respChan.C:
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+
+	pubAddrs := &PubAddrs{}
+	err = json.Unmarshal(msg.Data, pubAddrs)
+	if err != nil {
+		return nil, err
+	}
+
+	return pubAddrs, nil
+}
+
+func (c *TunaSessionClient) dialAConn(remoteAddr string, sessionID []byte, i int, addr *PubAddr, config *nkn.DialConfig) (conn *Conn, err error) {
+	if addr == nil || len(addr.IP) == 0 || addr.Port == 0 {
+		return nil, fmt.Errorf("dialAConn wrong params addr")
+	}
+
+	ctx := context.Background()
+	var cancel context.CancelFunc
+	if config.DialTimeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(config.DialTimeout)*time.Millisecond)
+		defer cancel()
+	}
+
+	dialer := &net.Dialer{}
+	netConn, err := dialer.DialContext(ctx, "tcp", fmt.Sprintf("%s:%d", addr.IP, addr.Port))
+	if err != nil {
+		return nil, err
+	}
+	conn = newConn(netConn)
+
+	err = writeMessage(conn, []byte(c.addr.String()), time.Duration(config.DialTimeout)*time.Millisecond)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	metadata := &pb.SessionMetadata{
+		Id: sessionID,
+	}
+	metadataRaw, err := proto.Marshal(metadata)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	buf, err := c.encode(metadataRaw, remoteAddr)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	err = writeMessage(conn, buf, time.Duration(config.DialTimeout)*time.Millisecond)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	sessKey := sessionKey(remoteAddr, sessionID)
```

Comment:
This piece of code is duplicated (and might conflict with) L543

Reply:
I suggest deleting L543, to keep the initial dial and re-dial connection with the same logic. How is it?